### ISSUE TITLE
[1.96] Custom extension with "self signed certificate in certificate chain" since 1.96

### ISSUE THREAD

**Issue by stannie42** [microsoft/vscode#236091]
Does this issue occur when all extensions are disabled?: Yes/No

Hello,

we've developed a custom extension and since we've upgraded to VS Code 1.96.0, we hit a "failed, reason: self signed certificate in certificate chain.". This error happens only in laptop that are using a proxy configuration and with this specific version of VS Code (1.96). When they rollback to 1.95, everything goes fine. Our extension just fetches our an API endpoint exposed with a trusted self signed certificate (with public certificate being added in local OS keystore/truststore).

We suspect this change to be the cause of our issue: https://github.com/microsoft/vscode/commit/3d41ba214ceb70ea8fc0fce6a977546aacbad2cd#diff-c5f10117e4d60a701184e817193c77b097be82b8708494ebc44153b7c13b4731R170

Do you have an idea how we should fix rapidly our issue? Thank you very much.

- Version: 1.96.0
- Commit: 138f619c86f1199955d53b4166bef66ef252935c
- Date: 2024-12-11T02:29:09.626Z
- Electron: 32.2.6
- ElectronBuildId: 10629634
- Chromium: 128.0.6613.186
- Node.js: 20.18.1
- V8: 12.8.374.38-electron.0
- OS: Windows_NT x64 10.0.19045

Steps to Reproduce:

1. Configure proxy / certificate to trust on your local machine
2. Install VS Code 1.96
3. Install our custom extension
4. Trigger a request against an API endpoint using a self signed certificate.

**Comment by chrmarti** [comment]
@stannie42 Could you share the code you are using to run `fetch`?

**Comment by stannie42** [comment]
@chrmarti sure, we run this code to fetch: https://github.com/continuedev/continue/blob/9c1132520953a83d40f654a0a483cacf763804da/core/llm/index.ts#L342

**Comment by chrmarti** [comment]
Does it work when you disable the following user setting?
<img width="483" alt="Image" src="https://github.com/user-attachments/assets/2b94d86a-40f6-461e-84d3-de80991af569" />

**Comment by mhelleboid** [comment]
It does not work with or without this setting

**Comment by mhelleboid** [comment]
is there an easy way to have a vscode 1.96 with vscode-proxy-agent 0.23?

**Comment by stannie42** [comment]
No, tried several times with / and without ticking this setting and same behaviour :(

Managed to get it working with "Network Proxy Test" extension but it looks like it loads the system certificate directly in code with this extension.

**Comment by chrmarti** [comment]
This might be from a different change then. Are you setting the certificates in the request options or the agent?

It would help to know which change caused it. We maintain a node.js based CLI tool `vscode-bisect` that automatically downloads and runs previous VS Code insiders and asks for each build if the issue reproduces. It takes up to 8 steps to find exactly the build that caused the regression.

Steps:
- install [node.js](https://nodejs.org/en/download) (if not yet done)
- npx --yes @vscode/vscode-bisect@latest
- follow the instructions until you found the offending build (its fine to leave the commit empty)
- report back the commit range in this issue

Thanks!

**Comment by mhelleboid** [comment]
It seems that our custom root certificate is not loaded anymore in VSCode 1.96.
By adding it manually in the agent used by our fetch it, we can workaround the problem from our extension for now.

Thanks for the information about vscode-bisect, we'll try and keep you posted

**Comment by chrmarti** [comment]
It seems by using `node-fetch` you end up using the `https` module underneath.

Two recent changes come to mind:
- A bug fix where we no longer add the system certificates when the agent options have `ca` set. (https://github.com/microsoft/vscode-proxy-agent/compare/ec30df6ca4347fd62d6a511ff74d524ea0e10f55..3d209d2e298898a63e2515f80b2c4aa5c339c1b7)
- A bug fix where we now add the system certificates to the proxy agent's options. (https://github.com/microsoft/vscode-proxy-agent/commit/f09ed5dd26afdcfdd182c7a11c1ad589925ca914)

If you previously added certificates to the request and relied on the system certificates being added to that, the first fix would potentially break it. This fix brings the behavior back in line with the Node.js documentation and avoids the potential for weakened security due to more certificates being accepted than passed in with the agent options.

I'm not sure I have the full picture though, could you also share a code pointer to where you assemble the request options?

**Comment by stannie42** [comment]
quickly replying with code fetch: https://github.com/continuedev/continue/blob/main/packages/fetch/src/fetch.ts

**Comment by chrmarti** [comment]
That looks like you are affected by that fix. I think the fix is still good because it aligns the behavior with Node.js, but it unfortunately breaks this particular setup.

Adding your custom certificate to the request makes sense. I assume it would be part of the system certificates (registered in the OS keychain)?

Alternatively: To get `tls.rootCertificates` and the system certificates automatically set, you could try omitting the `ca` option for the proxy agents. VS Code then adds these certificates.

**Comment by TheMasterFX** [comment]
I have the some problem. We use VSCode a different AI extensions (like Continue). Our OpenAI API endpoint is self hosted to keep the data in our company. This is a https connection with a certificate signed by our company Root CA. Since 1.96 we can't use the AI extensions anymore.

**Comment by gjsjohnmurray** [comment]
@TheMasterFX have you raised the issue with the extension author?

**Comment by TheMasterFX** [comment]
> [@TheMasterFX](https://github.com/TheMasterFX) have you raised the issue with the extension author?

There are already some issues about that, like https://github.com/continuedev/continue/issues/1423
But it seems like nothing happened so far. I didn't check every extension.
But in my understanding, it should be better if all the certificates in the Windows/Linux Store should be passed to the extensions, so it must only implemented once and also make it easier to IT departments to provide the company certificates via Active Directory.

**Comment by chrmarti** [comment]
@TheMasterFX We already add all root certificates from the OS to the networking for extensions. https://github.com/continuedev/continue/issues/1423 predates the changes discussed above. Could you share the code with which you see this problem?
